miner/block.go

```diff
@@ -0,0 +1,397 @@
+// Copyright 2021 The celo Authors
+// This file is part of the celo library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package miner
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/consensus"
+	"github.com/celo-org/celo-blockchain/contract_comm/random"
+	"github.com/celo-org/celo-blockchain/core"
+	"github.com/celo-org/celo-blockchain/core/rawdb"
+	"github.com/celo-org/celo-blockchain/core/state"
+	"github.com/celo-org/celo-blockchain/core/types"
+	"github.com/celo-org/celo-blockchain/log"
+	"github.com/celo-org/celo-blockchain/params"
+)
+
+// blockState is the collection of modified state that is used to assemble a block
+type blockState struct {
+	signer types.Signer
+
+	state    *state.StateDB // apply state changes here
+	tcount   int            // tx count in cycle
+	gasPool  *core.GasPool  // available gas used to pack transactions
+	gasLimit uint64
+
+	header         *types.Header
+	txs            []*types.Transaction
+	receipts       []*types.Receipt
+	randomness     *types.Randomness // The types.Randomness of the last block by mined by this worker.
+	txFeeRecipient common.Address
+}
+
+// constructAndSubmitNewBlock constructs a new block and if the worker is running, submits
+// a task to the engine
+func (w *worker) constructAndSubmitNewBlock(ctx context.Context) {
+	start := time.Now()
+
+	// Initialize the block.
+	b, err := w.prepareBlock()
+	if err != nil {
+		log.Error("Failed to create mining context", "err", err)
+		return
+	}
+	w.updateSnapshot(b)
+
+	// TODO: worker based adaptive sleep with this delay
+	// wait for the timestamp of header, use this to adjust the block period
+	delay := time.Until(time.Unix(int64(b.header.Time), 0))
+	select {
+	case <-time.After(delay):
+	case <-ctx.Done():
+		return
+	}
+
+	err = b.selectAndApplyTransactions(ctx, w)
+	if err != nil {
+		log.Error("Failed to apply transactions to the block", "err", err)
+		return
+	}
+	w.updateSnapshot(b)
+
+	block, err := b.finalizeAndAssemble(w)
+	if err != nil {
+		log.Error("Failed to finalize and assemble the block", "err", err)
+		return
+	}
+	w.updateSnapshot(b)
+	if w.isRunning() {
```

Comment:
we could change `constructAndSubmitNewBlock` to `createNewBlock() blockInfo`. So, it will be responsible to create a new block, add txs to it, and will return the create block (plus state, receipt, etc) as a result.

That way, it's less couple with the overall worker, doesn't need to know about channel, and the overall block flow.

Then caller will then get the result and submit it to the channel.

That way, the `pipeline` is defined on a single place:

```go
func blockPipeline(ctx) {
  block, receipts, state, created := createNewBlock(ctx)
  sealedBlock, err := engine.Seal(ctx, task.block)
  // result loop code
  addBlockToChain()
}
```

Reply:
You need to split the pipeline into the top half and the bottom half [(creation & seal) an (result loop & insertion)] as the result that you get back from the results loop channel does not have to be associated with the task that the specific instance of the pipeline submitted. As such I don't want to mix up state by working on different tasks in the same local (function) context.